### PR TITLE
search for the cargo toml based on the current file

### PR DESCRIPTION
**Pull request by jannickj**
Code for the proposed fix: #222

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -53,6 +54,10 @@ function didOpenTextDocument(document: TextDocument, context: ExtensionContext):
         return;
     }
     folder = getOuterMostWorkspaceFolder(folder);
+    folder = getCargoTomlWorkspace(folder, document.uri.fsPath);
+    if (!folder) {
+        return;
```
Comment (nrc) [inline src/extension.ts]: Can/should we issue a warning about missing Cargo.toml here?

Comment (jannickj) [reply src/extension.ts]: yes probably a good idea

Comment (jannickj) [reply src/extension.ts]: My only concern is that you will get a pop up if just opening a .rs file outside it's project

**Comment by jannickj** [comment]
Changed how the message is shown from a pop up to a nice little message in the bottom, much more pleasant from a UX perspective.

File: src/extension.ts (hunk 1 above)
Comment (jannickj) [reply src/extension.ts]: Okay I've come up with a solution that will work in both use cases.

**Comment by maelvalais** [comment]
Thank you, @jannickj!!!  🎉 🍾

**Comment by rickvanprim** [comment]
@nrc @jannickj I see this changed was backed out.  Any context on why?  Will this feature be re-added soon?

**Comment by Xanewok** [comment]
IIRC this introduced a bug that broke project detection and it did impact a lot of users.

Could you remind me what this aims to fix? Not warning against missing Cargo.toml when the user opens a project that has a Rust project nested and not at the top level?

I'll try to tackle this again in the immediate future.

**Comment by rickvanprim** [comment]
I can't speak for others, but for me this being able to have a project that contains one or more Rust crates mixed in with other projects, where Cargo is not the root build system and Rust is not the only language.  RLS detecting Cargo.toml relative to the source file being looked at means the crate can be located anywhere in the repo, not just at the root.

**Comment by rickvanprim** [comment]
@Xanewok any updates?

**Comment by jannickj** [comment]
@Xanewok The problem is with having multiple rust project in the space workspace.

The current setup is that the `Cargo.toml` must be in the root of the workspace. However with this change it would allow the project to work in any subfolder.

@rickvanprim I remember the problem being impact on windows users. I use linux exclusively and hadn't properly tested nor forwarned that it probably should be tested.

I will see if I can fix the compatibility problems, and then try again and hopefully we can have others help the testing.

**Pull request merged**